### PR TITLE
version 1: identity path

### PR DESCRIPTION
**Pull request by jreiberkyle**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,41 @@
...
+# Radiometric Normalization #
+
+This library contains the functions to analyse a set of reference images for features that have a stable level of intensity over the whole reference set and use these features to normalize the intensity distribution of separate candidate image.
+
+It was originally created to radiometrically normalize a satellite image to a time series from a reference
+(atmospherically corrected) dataset.
+
+## How to use ##
+
+This git repo contains a vagrant environment within which to run the library in. To start the vagrant environment, in the root directory of the repo type:
```
Comment (MattFerraro) [inline README.md]: "a vagrant environment to run the library in" or "a vagrant environment within which to run the library"

File: README.md [hunk 2]
```diff
@@ -1,2 +1,41 @@
...
+If successful it should end with:
+
+    vagrant@precise64:/vagrant$ nosetests ./tests 
+    .............
+    ----------------------------------------------------------------------
+    Ran 13 tests in 0.112s
+
+    OK
+
+## Organisation of the repo ##
```
Comment (MattFerraro) [inline README.md]: hmm, I'd say probably stick with american english because we're an american company

File: init.sh [hunk 3]
```diff
@@ -0,0 +1,28 @@
+#!/bin/bash
+
+if [[ $EUID -eq 0 ]]; then
+  echo "Do *not* run this as root, sudo will be used as needed."
```
Comment (MattFerraro) [inline init.sh:4]: nice!

File: radiometric_normalization/README.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+* Each image should should cover exactly the same geographic location
+* Each image should contain exactly the same number of pixels (rows and cols)
+* Each image should have the same no data value
+
+### Function ###
+
+This module analyses the set of images and creates a single output image that reflects the set of input images (e.g. the mean of each pixel over time) as well as a weighting alpha mask that gives an indication of how strong each pixel would be if used as a pseudo-invariant feature.
+
+### Algorithm ###
+* Mean with uniform weight: Currently only this method is implemented. It calculates the mean of each pixel over time and has a uniform weight (0 for pixels where it is a no data pixel for every image in the time stack) and 65535 for valid pixels.
```
Comment (MattFerraro) [inline radiometric_normalization/README.md:26]: This level of writeup in a readme is excellent. It's the perfect amount that I could reasonably use the modules without ever having to look at their implementation

File: radiometric_normalization/gimage.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+# In-memory geospatial image
+# Bands: list of uint16 numpy arrays, each holding a band of data
+# Alpha: uint16 numpy array holding the alpha information
+# Metadata: dict containing georeferencing information
+# (geotransform, projection, rpc)
+GImage = namedtuple('GImage', 'bands, alpha, metadata')
+
+
+def save(gimage, filename):
```
Comment (MattFerraro) [inline radiometric_normalization/gimage.py]: It would be nice to expose a save_to_file(gimage, file_like_object) function so that clients can write to in-memory files if they desire. Then this function could be implemented as:

```
def save(gimage, filename):
    with open(filename, "w') as file_object:
        save_to_file(gimage, file_object)
```

This would become important if we wanted to do something like stand up a flask server that generated these types of images on the fly and served them up to users. If the file got written to disk, we could serve it no problem but the http request would terminate immediately after serving the file and we'd have no way to clean it up afterwards!

Using a file-like object means we could instead use python's built-in TemoraryFile mechanisms which are destroyed once they go out of scope.

File: radiometric_normalization/gimage.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+        options.append('PHOTOMETRIC=RGB')
+
+    datatype = gdal_array.NumericTypeCodeToGDALTypeCode(
+        gimage.bands[0].dtype.type)
+    ysize, xsize = gimage.bands[0].shape
+    gdal_ds = gdal.GetDriverByName('GTIFF').Create(
+        filename, xsize, ysize, band_count, datatype,
+        options=options)
+
+    for i in range(len(gimage.bands)):
```
Comment (MattFerraro) [inline radiometric_normalization/gimage.py]: Consider using enumerate:

```
for i, band in enumerate(gimage.bands):
    gdal_array.BandWriteArray(gdal_ds.GetRasterBand(i + 1), band)
```

File: radiometric_normalization/normalize.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    def apply_lut(band, lut):
+        'Changes band intensity values based on intensity look up table (lut)'
+        if lut.dtype != band.dtype:
+            raise Exception(
+                "Band ({}) and lut ({}) must be the same data type.").format(
+                band.dtype, lut.dtype)
+        return numpy.take(lut, band, mode='clip')
+
+    img = gimage.load(input_path)
+    for i in range(len(img.bands)):
```
Comment (MattFerraro) [inline radiometric_normalization/normalize.py:51]: same here, maybe use enumerate

File: radiometric_normalization/pif.py [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+
+from osgeo import gdal
+
+
+def generate(candidate_path, time_stack_path='time_stack.tif',
+             method='identity'):
+    ''' Generates psuedo invariant features as a list of pixel pairs
+
+    Input:
+        time_stack_path (str): A path to the time stack image
```
Comment (MattFerraro) [inline radiometric_normalization/pif.py]: could you add a description of "candidate_path"?

File: radiometric_normalization/pif.py [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+    assert no_bands_reference == no_bands_candidate, \
+        '%s and %s have different number of bands' % \
+        (candidate_path, time_stack_path)
+    assert candidate_ds.RasterXSize == time_stack_ds.RasterXSize and \
+        candidate_ds.RasterYSize == time_stack_ds.RasterYSize, \
+        '%s and %s have different sizes' % \
+        (candidate_path, time_stack_path)
+
+    if method is 'identity':
```
Comment (MattFerraro) [inline radiometric_normalization/pif.py]: probably want to use == instead of "is" here

File: radiometric_normalization/pif.py [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+        pixel_pairs = _filter_nodata(time_stack, weighting,
+                                     candidate, candidate_mask)
```
Comment (MattFerraro) [inline radiometric_normalization/pif.py]: should there be something like

```
if method == 'identity':
    stuff()
else:
    raise NotImplementedError("I don't know this mode!")
```

?

File: radiometric_normalization/time_stack.py [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+    - Since the output file is only meant to be used internally, the
+        geographic information isn't carried over to the output file.
+
+    Input:
+        image_paths (list of str): A list of the input file paths
+        method (str): Time stack creation method [identity]
+        output_path (str): A path to write the file to
+    '''
+
+    output_nodata = 2 ** 15 - 1
```
Comment (MattFerraro) [inline radiometric_normalization/time_stack.py]: I like this way of writing constants, I think it's clearer than 32767

File: radiometric_normalization/time_stack.py [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+    output_datatype = numpy.uint16
+
+    all_bands = _read_in_bands(image_paths)
+    if method is 'identity':
```
Comment (MattFerraro) [inline radiometric_normalization/time_stack.py]: == is probably better than "is"

File: radiometric_normalization/time_stack.py [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+    Output:
+        all_bands (list of list of arrays): A list of each band,
+            each band consisting of a list of images each entry being a
+            masked array of the image data
+    '''
+
+    no_images = len(all_images)
+    no_bands, _, _ = all_images[0].shape
+    all_bands = []
+    for band in range(no_bands):
```
Comment (MattFerraro) [inline radiometric_normalization/time_stack.py]: This loop is a little hard to parse. Maybe just a simple change like

```
for band_index in range(no_bands):
```

File: tests/pif_tests.py [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+                                              [nodata, 4]],
+                                             dtype='uint16'),
+                                 numpy.array([[3, 3],
+                                              [nodata, 1]],
+                                             dtype='uint16')])
+        candidate_mask = numpy.array([[65535, 65535],
+                                      [0, 65535]],
+                                     dtype='uint16')
+
+        golden_output_list = [{'coordinates': (0, 1),
```
Comment (MattFerraro) [inline tests/pif_tests.py:38]: This is an excellent test!

File: tests/time_stack_tests.py [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+        rows, cols = all_bands[0][0].shape
+
+        self.assertEqual(no_bands, 5)
+        self.assertEqual(no_images, 7)
+        self.assertEqual(rows, 10)
+        self.assertEqual(cols, 11)
+        for image in range(no_images):
+            for band in range(no_bands):
+                for row in range(rows):
+                    for col in range(cols):
```
Comment (MattFerraro) [inline tests/time_stack_tests.py]: With this many levels of nesting, it's probably not a bad idea to use xrange instead of range

File: tests/time_stack_tests.py [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+
+from radiometric_normalization import time_stack
+
+
+class Tests(unittest.TestCase):
+    def test_organise_images_to_bands(self):
+        # Seven images of five bands of 10 by 11 pixels
+        all_images = []
+        nodata = 2 ** 15 - 1
+        for i in range(7):
```
Comment (MattFerraro) [inline tests/time_stack_tests.py]: general convention is to use xrange for something like this, but it's not a big deal. I appreciate that you choose one option (range) and stick with it

File: tests/transformation_tests.py [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+class Tests(unittest.TestCase):
+    def test_get_transform_function(self):
+        self.assertEqual(
+            transformation.get_transform_function('linear_relationship'),
+            transformation.linear_relationship)
+        self.assertRaises(Exception,
+                          transformation.get_transform_function,
+                          'unknown')
+
+    def test_pifs_to_pifset(self):
```
Comment (MattFerraro) [inline tests/transformation_tests.py:31]: lol I find the name "pifset" really confusing because I'd expect it to just be a [set](https://jobs.planet-labs.com/v0/jobs/25731632) of pifs!

**Comment by MattFerraro** [comment]
Overall this looks great! I don't think I'll ever get used to how foreign and unpythonic numpy code looks, but other than that the interfaces, documentation, and overall style seem spot on to me. It might be nice to move away from using the file system so much and let clients handle the raw OS stuff their own way, but that's an imagined use case not a real one.

File: radiometric_normalization/gimage.py (hunk 5 above)
Comment (jreiberkyle) [inline radiometric_normalization/gimage.py]: This is a great idea. I implemented a tweaked version to accommodate the gdal paradigm:

```
def save(gimage, filename):
    gdal_ds = create_ds(gimage, filename)
    save_to_ds(gimage, gdal_ds)
```

`create_ds` creates the output gdal dataset (similar to file_object) and `save_to_ds` writes the gimage to the gdal_ds. One gotcha here is that gdal_ds is a pointer and it must be set to None for the file to be generated. Example:

```
output_ds = gdal.GetDriverByName('GTiff').Create(
            output_file, 2, 2, 2, gdal.GDT_UInt16,
            options=['PHOTOMETRIC=RGB', 'ALPHA=YES'])

gimage.save_to_ds(test_gimage, output_ds)
output_ds = None
``
```

File: radiometric_normalization/pif.py [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+        method (str): Which psuedo invariant feature generation method to use
+
+    Output:
+        pixel_pairs (list of pixel pair dict): Pixel pairs are
+            {'coordinates', (int, int),
+             'weighting', float,
+             'reference', list of numbers,
+             'candidate', list of numbers}
+    '''
+    time_stack_ds = gdal.Open(time_stack_path)
```
Comment (jreiberkyle) [inline radiometric_normalization/pif.py]: gimage.load() should be used to load time_stack and candidate images

File: radiometric_normalization/pif.py [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+        (candidate_path, time_stack_path)
+
+    if method is 'identity':
+        pixel_pairs = _filter_nodata(time_stack, weighting,
+                                     candidate, candidate_mask)
+
+    return pixel_pairs
+
+
+def _filter_nodata(time_stack, weighting, candidate, candidate_mask):
```
Comment (jreiberkyle) [inline radiometric_normalization/pif.py]: the name _filter_nodata and the implementation (creates pixel pairs) don't match up

File: radiometric_normalization/time_stack.py [hunk 20]
```diff
@@ -0,0 +1,214 @@
...
+        alpha_band = image_ds.GetRasterBand(image_ds.RasterCount)
+        assert alpha_band.GetColorInterpretation() != gdal.GCI_AlphaBand, \
+            '%s has an alpha band' % image_path
+
+        all_images.append(image_ds.ReadAsArray())
+
+    return all_images, nodata
+
+
+def _organise_images_to_bands(all_images, nodata):
```
Comment (jreiberkyle) [inline radiometric_normalization/time_stack.py]: change organise to organize

File: radiometric_normalization/time_stack.py [hunk 21]
```diff
@@ -0,0 +1,214 @@
...
+        all_bands_mask = numpy.logical_or(band_mask, all_bands_mask)
+        output_bands.append(band_mean.astype(output_datatype))
+
+    output_mask = numpy.logical_not(all_bands_mask).astype(output_datatype) * \
+        numpy.iinfo(output_datatype).max
+
+    return output_bands, output_mask
+
+
+def _write_out_bands(output_bands, mask,
```
Comment (jreiberkyle) [inline radiometric_normalization/time_stack.py]: use gimage.save instead

File: radiometric_normalization/time_stack.py [hunk 22]
```diff
@@ -0,0 +1,214 @@
...
+    gdal_ds = gdal.GetDriverByName('GTIFF').Create(
+        output_path, cols, rows, no_bands + 1, datatype,
+        options=options)
+
+    # Write output file data
+    for i in range(no_bands):
+        gdal_array.BandWriteArray(
+            gdal_ds.GetRasterBand(i + 1),
+            output_bands[i])
+    gdal_ds.GetRasterBand(1).SetNoDataValue(output_nodata)
```
Comment (jreiberkyle) [inline radiometric_normalization/time_stack.py]: I believe the nodata value here should be set to the same as the input_nodata value, otherwise nodata pixels won't be stored correctly

File: radiometric_normalization/time_stack.py [hunk 23]
```diff
@@ -0,0 +1,214 @@
...
+    rows = test_image_ds.RasterYSize
+    nodata = test_image_ds.GetRasterBand(1).GetNoDataValue()
+
+    # Read in each image as a numpy array
+    all_images = []
+    for image_path in image_paths:
+        image_ds = gdal.Open(image_path)
+
+        logging.info('Processing ' + image_path + '...')
+        assert image_ds.RasterCount == no_bands, \
```
Comment (jreiberkyle) [inline radiometric_normalization/time_stack.py]: use "".format() to format strings

File: tests/pif_tests.py [hunk 24]
```diff
@@ -0,0 +1,77 @@
...
+limitations under the License.
+'''
+import unittest
+import numpy
+
+from radiometric_normalization import pif
+
+
+class Tests(unittest.TestCase):
+    def test_create_pif_dict(self):
```
Comment (jreiberkyle) [inline tests/pif_tests.py]: add unit tests for the other functions in pif.py

File: tests/time_stack_tests.py [hunk 25]
```diff
@@ -0,0 +1,137 @@
...
+import unittest
+import numpy
+import os
+
+from osgeo import gdal
+
+from radiometric_normalization import time_stack
+
+
+class Tests(unittest.TestCase):
```
Comment (jreiberkyle) [inline tests/time_stack_tests.py:22]: add tests for all time_stack functions

File: radiometric_normalization/pif.py [hunk 26]
```diff
@@ -0,0 +1,102 @@
...
+            image (0 is nodata)
+
+    Output:
+        pixel_pairs (list of pixel pair dict): Pixel pairs are
+            {'coordinates', (int, int),
+             'weighting', float,
+             'reference', list of numbers,
+             'candidate', list of numbers}
+    '''
+    all_mask = numpy.logical_not(numpy.logical_or(
```
Comment (jreiberkyle) [inline radiometric_normalization/pif.py:66]: appears to be filtering based on mask, not on pixel value == nodata

File: radiometric_normalization/time_stack.py [hunk 27]
```diff
@@ -0,0 +1,214 @@
...
+            masked array of the image data
+    '''
+
+    all_images, nodata = _read_in_gdal_images(image_paths)
+    all_bands = _organise_images_to_bands(all_images, nodata)
+
+    return all_bands
+
+
+def _read_in_gdal_images(image_paths):
```
Comment (jreiberkyle) [inline radiometric_normalization/time_stack.py]: Use gimage to load images. This changes the data structure so _read_in_bands will also change

File: radiometric_normalization/time_stack.py [hunk 28]
```diff
@@ -0,0 +1,214 @@
...
+
+        logging.info('Processing ' + image_path + '...')
+        assert image_ds.RasterCount == no_bands, \
+            '%s has a different number of bands' % image_path
+        assert image_ds.RasterXSize == cols and image_ds.RasterYSize == rows, \
+            '%s has a different size' % image_path
+        assert image_ds.GetRasterBand(1).GetNoDataValue() == nodata, \
+            '%s has a different nodata value' % image_path
+        alpha_band = image_ds.GetRasterBand(image_ds.RasterCount)
+        assert alpha_band.GetColorInterpretation() != gdal.GCI_AlphaBand, \
```
Comment (jreiberkyle) [inline radiometric_normalization/time_stack.py]: Why are images with alpha bands not supported? We could use the alpha band in the weighting calculation

**Comment by arincos** [comment]
Changes implemented in v1_identity_path_code_review (Pull request https://github.com/planetlabs/radiometric_normalization/pull/3 )
